File: gaming（框架代码）/gaming-design/src/main/java/org/gaming/design/loader/DesignCache.java
/**
 * 
 */
package org.gaming.design.loader;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.gaming.design.exception.DesignCacheLoadingException;
import org.gaming.design.exception.DesignNotFoundException;
import org.gaming.tool.GsonUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author YY
 *
 */
public abstract class DesignCache<T extends IDesignData> {

	protected static Logger logger = LoggerFactory.getLogger(DesignCache.class);
	
	private List<T> dataList;
	private Map<Integer, T> dataMap;
	
	private Class<T> clazz;
	private boolean hasError = false;
	
	private DesignParser designParser;
	
	@SuppressWarnings("unchecked")
	public DesignCache() {
		this.clazz = (Class<T>) ((ParameterizedType) this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
		DesignCacheManager.register(this);
		this.designParser = new JsonDesignParser();
	}
	
	public abstract String fileName();
	
	protected void loadAutoGenerate() {}
	protected void loadAfterReady() {}
	protected void loadAfterAllReady() {}
	protected boolean check() {
		return true;
	}
	
	public final void load(List<DesignField> fields, List<String[]> bodyDatas) throws DesignCacheLoadingException {
		List<T> dataList = new ArrayList<>();
		Map<String, Field> fieldMap = getDeclaredFields();
		for(int i = 0; i < bodyDatas.size(); i++) {
			String[] data = bodyDatas.get(i);
			T obj = formObject(i, fieldMap, fields, data);
			dataList.add(obj);
		}
		load(dataList);
	}
	
	public final void load(List<T> dataList) throws DesignCacheLoadingException {
		Map<Integer, T> dataMap = new HashMap<>();
		
		hasError = false;
		for(T obj : dataList) {
			if(dataMap.containsKey(obj.id())) {
				logger.error("文件【{}】在解释时发现的重复ID【{}】的数据", fileName(), obj.id());
				hasError = true;
				continue;
			}
			dataMap.put(obj.id(), obj);
		}
		this.dataList = dataList;
		this.dataMap = dataMap;
		
		loadAutoGenerate();
		loadAfterReady();
		
		if(hasError) {
			throw new DesignCacheLoadingException(fileName());
		}
	}
	
	public final List<T> all() {
		return this.dataList;
	}
	
	public final T getOrNull(int key) {
		return dataMap.get(key);
	}
	
	public final T getOrThrow(int key) {
		T t = dataMap.get(key);
		if(t == null) {
			throw new DesignNotFoundException(this.getClass().getSimpleName() + ".getOrThrow", key);
		}
		return t;
	}
	
	private Map<String, Field> getDeclaredFields() {
		Map<String, Field> fieldMap = new HashMap<>();
		for(Field field : clazz.getDeclaredFields()) {
			fieldMap.put(field.getName(), field);
		}
		//超类的字段
		Class<?> superClass = clazz.getSuperclass();
		while(superClass != null) {
			for(Field field : superClass.getDeclaredFields()) {
				fieldMap.put(field.getName(), field);
			}
			superClass = superClass.getSuperclass();
		}
		return fieldMap;
	}
	
	/**
     * 创建对象，并进行自动赋值
     *
     * @param rowNum  行号
     * @param rowData
     * @return
     */
    private final T formObject(int rowIndex, Map<String, Field> fieldMap, List<DesignField> fields, String[] bodyData) {
    	Map<String, String> headToBody = new HashMap<>();
    	for (int i = 0; i < fields.size() && i < bodyData.length; i++) {
    		headToBody.put(fields.get(i).getName(), bodyData[i]);
    	}
		T obj = newInstance();
    	for(Entry<String, String> entry : headToBody.entrySet()) {
    		Field field = fieldMap.get(entry.getKey());
    		if(field == null) {
    			continue;
    		}
    		String bodyValue = entry.getValue();
    		try {
				Object value = formValue(field, bodyValue);
				if (value == null) {
					continue;
				}
				field.setAccessible(true);
				field.set(obj, value);
				field.setAccessible(false);
			} catch (Exception e) {
				logger.error("文件【{}】在解释第【{}】行字段【{}】的数据时发现错误，无法解释的数据【{}】", fileName(), rowIndex, field.getName(), bodyValue);
				logger.error(e.getMessage());
				hasError = true;
			}
    	}
        return obj;
    }
    
    private T newInstance() {
    	try {
			T obj = clazz.newInstance();
			return obj;
		} catch (InstantiationException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		}
    	return null;
    }
	
    
    private Object formValue(Field field, String value) {
    	Object valueObj = null;
		
		if(int.class.equals(field.getType()) || Integer.class.equals(field.getType())) {
			
			valueObj = Integer.parseInt(value);
			
		} else if(long.class.equals(field.getType()) || Long.class.equals(field.getType())) {
			
			valueObj = Long.parseLong(value);
			
		} else if(float.class.equals(field.getType()) || Float.class.equals(field.getType())) {
			
			valueObj = Float.parseFloat(value);
			
		} else if(String.class.equals(field.getType())) {
			
			valueObj = value;
			
		} else if(List.class.isAssignableFrom(field.getType())) {
			
			Class<?> paramClass = (Class<?>) ((ParameterizedType) field.getGenericType())
					.getActualTypeArguments()[0];
			valueObj = designParser.parseList(value, paramClass);
			
		} else if(Map.class.isAssignableFrom(field.getType())) {
			
			Class<?> keyClazz = (Class<?>) ((ParameterizedType) field.getGenericType())
					.getActualTypeArguments()[0];
			Class<?> valueClazz = (Class<?>) ((ParameterizedType) field.getGenericType())
					.getActualTypeArguments()[0];
			
			valueObj = designParser.parseMap(value, keyClazz, valueClazz);
			
		} else {
			valueObj = designParser.parseStruct(value, field.getType());
			
		}
		return valueObj;
    }
    
    protected final void loadConst(Map<String, String> constMap) {
    	for(Field field : this.getClass().getSuperclass().getDeclaredFields()) {
    		String constValue = constMap.get(field.getName());
    		if(constValue == null) {
    			logger.error("{}缺少索引名为{}的常量配置", fileName(), field.getName());
    			continue;
    		}
    		Object value = formValue(field, constValue);
			if (value == null) {
				continue;
			}
			try {
				field.setAccessible(true);
				field.set(this, value);
				field.setAccessible(false);
			} catch (IllegalArgumentException e) {
				e.printStackTrace();
			} catch (IllegalAccessException e) {
				e.printStackTrace();
			}
    	}
    }
    
    /**
     * 打印当前缓存中的数据
     * @return
     */
    public String printCacheData() {
    	StringBuilder builder = new StringBuilder();
    	List<Field> fieldList = new ArrayList<>();
    	for(Field field : clazz.getDeclaredFields()) {
    		builder.append(field.getName()).append("\t");
    		field.setAccessible(true);
    		fieldList.add(field);
    	}
    	try {
	    	for(T t : this.dataList) {
	    		builder.append("\n");
	    		for(Field field : fieldList) {
					builder.append(GsonUtil.toJson(field.get(t))).append("\t");
	    		}
	    	}
	    	System.out.println(builder.toString());
    	} catch (IllegalArgumentException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		}
    	return builder.toString();
    }

	public DesignParser getDesignParser() {
		return designParser;
	}

	public void setDesignParser(DesignParser designParser) {
		this.designParser = designParser;
	}
}

File: gaming（框架代码）/gaming-design/typescript-code/DesignCache.ts
import DesignDataNotFoundError from "./DesignDataNotFoundError";
import { SeparatorDesignParser } from "./SeparatorDesignParser";

export default abstract class DesignCache<T extends IDesignData> {

    private list: T[];
    private map: Map<number, T>;
	private parser: SeparatorDesignParser;
	/**
	 * 后续实现
	 */
    public loadData(text: string): void {
        let list: T[] = [];
        let datas: object[] = JSON.parse(text);
        for(let i = 0; i < datas.length; i++) {
            let data: object = datas[i];
            console.log(data);
        }
    }
	
	/**
	 * 通过原数据与分隔符解析器加载配置对象
	 */
    public loadObjects(datas: object[], parser: SeparatorDesignParser): void {
    	this.parser = parser;
        const list: T[] = [];
        const map: Map<number, T> = new Map<number, T>();
        try {
            for(let i = 0; i < datas.length; i++) {
                let temp: any = datas[i];
                let t: T = this.createInstance();
                for(const key in t) {
                    if(temp[key] === undefined) {
                        if(typeof t[key] === "number") {
                        } else if(typeof t[key] === "string") {
                        } else if(Array.isArray(t[key])) {
                        } else if(typeof t[key] === "object") {
                            let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                            let obj: any = t["format" + upperFirst]();
                            t[key] = this.parser.parseStruct("", obj);
                        }
                    } else {
                        if(typeof t[key] === "number") {
                            t[key] = temp[key];
                        } else if(typeof t[key] === "string") {
                            t[key] = temp[key];
                        } else if(Array.isArray(t[key])) {
                            let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                            let obj: any = t["format" + upperFirst]();
                            t[key] = this.parser.parseList(temp[key].toString(), obj);
                        } else if(typeof t[key] === "object") {
                            let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                            let obj: any = t["format" + upperFirst]();
                            t[key] = this.parser.parseStruct(temp[key].toString(), obj);
                        }
                    }
                }
                list.push(t);
                map.set(t.Id(), t);
            }
        } catch (error) {
            console.log(`加载${this.fileName}发生异常${error}`);
        }
        this.list = list;
        this.map = map;

        this.loadAutoGenerate();
    }

    protected abstract get fileName(): string;
    //创建对象
    protected abstract createInstance(): T;
    //加载自动生成的索引结构
    protected abstract loadAutoGenerate(): void;

    public all(): T[] {
        return this.list;
    }

    public getOrNull(id: number): T {
        let t: T = this.map.get(id);
        if(t === undefined) {
            return null;
        }
        return t;
    }

    public getOrThrow(id: number): T {
        let t: T = this.map.get(id);
        if(t === undefined) {
            throw new DesignDataNotFoundError(this.fileName + ".getOrThrow", id);
        }
        return t;
    }
    
    /**
     * 加载常量
     */
    protected loadConst(map: Map<string, string>): void {
        let t: any = this;
    	for(const key in t) {
            if(key === "list" || key === "map" || key === "parser") {
                continue;
            }
            if(map.get(key) === undefined) {
                if(typeof t[key] === "number") {
                } else if(typeof t[key] === "string") {
                } else if(Array.isArray(t[key])) {
                } else if(typeof t[key] === "object") {
                    let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                    let obj: any = t["format" + upperFirst]();
                    t[key] = this.parser.parseStruct("", obj);
                }
            } else {
                if(typeof t[key] === "number") {
                    t[key] = parseInt(map.get(key));
                } else if(typeof t[key] === "string") {
                    t[key] = map.get(key);
                } else if(Array.isArray(t[key])) {
                    let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                    let obj: any = t["format" + upperFirst]();
                    t[key] = this.parser.parseList(map.get(key), obj);
                } else if(typeof t[key] === "object") {
                    let upperFirst: string = key.substring(0, 1).toUpperCase() + key.substring(1);
                    let obj: any = t["format" + upperFirst]();
                    t[key] = this.parser.parseStruct(map.get(key), obj);
                }
            }
        }
    }

}